lib/google/gax/api_call.rb

```diff
@@ -0,0 +1,135 @@
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'google/gax/api_call/retry_policy'
+require 'google/gax/call_options'
+require 'google/gax/errors'
+
+module Google
+  module Gax
+    class ApiCall
+      ##
+      # Converts an RPC call into an API call.
+      #
+      # In typical usage, `stub_method` will be a proc used to make an RPC
+      # request. This will mostly likely be a bound method from a request Stub
+      # used to make an RPC call.
+      #
+      # The result is created by applying a series of function decorators
+      # defined in this module to `stub_method`.
+      #
+      # The result is another proc which has the same signature as the original.
+      #
+      # @param stub_method [Proc] used to make a bare rpc call
+      # @param timeout [Numeric] client-side timeout for API calls
+      # @param metadata [Hash] request headers
+      # @param retry_policy [Hash] the settings for error retry, will be merged
+      #   to the {CallOptions#retry_policy} object if supported.
+      # @param params_extractor [Proc] extracts routing header params from the
+      #   request
+      # @param exception_transformer [Proc] if an API exception occurs this
+      #   transformer is given the original exception for custom processing
+      #   instead of raising the error directly
+      def initialize(stub_method, timeout: nil, metadata: nil,
+                     retry_policy: nil, params_extractor: nil,
+                     exception_transformer: nil)
+        @stub_method           = stub_method
+        @timeout               = timeout
+        @metadata              = metadata
+        @retry_policy          = retry_policy
+        @params_extractor      = params_extractor
+        @exception_transformer = exception_transformer
+      end
+
+      ##
+      # Invoke the API call.
+      #
+      # @param request [Object] The request object.
+      # @param options [CallOption, Hash] The options for making the API call.
+      # @param block [Proc] The proc to call when the API call is made.
+      #
+      def call(request, options: nil, &block)
+        options = init_call_options options
+
+        apply_params_extractor! request, options
+
+        deadline = calculate_deadline(options)
+
+        begin
+          op = @stub_method.call(request, deadline: deadline,
+                                          metadata: options.metadata,
+                                          return_op: true)
+          res = op.execute
+          yield op if block
+          res
+        rescue StandardError => error
+          if retry?(deadline)
+            retry if options.retry_policy.call(error)
+          end
+
+          if error.is_a? GRPC::BadStatus
+            error = Google::Gax.from_error(error).new('RPC failed')
```

Comment:
One of the changes in this implementation is that wrapped errors match the underlying GRPC error, instead of being a RetryError. But I think it makes more sense for the GaxError to match the BadStatus instead of a RetryError.